### ISSUE TITLE
WT Preview 1.18 breaks Terminal.Gui keyboard handling (NetDriver)

### ISSUE THREAD

**Issue by tig** [microsoft/terminal#15693]
### Windows Terminal version

1.18.1462.0

### Windows build number

10.0.22631.1972

### Other Software

Terminal.Gui release version https://github.com/gui-cs/Terminal.Gui

### Steps to reproduce

1. Clone Terminal.Gui `develop` branch
2. `dotnet run --project .\UICatalog\ -usc` (-usc flag forces the NetConsole driver which uses the .NET Console API)
3. When UI Catalog appears, press CTRL-A to show About box

### Expected Behavior

Keyboard input should work. CTRL-A should show the About box in the UI Catalog app. CTRL-Q should quit.

If you do the same thing in WT v1.17 the keyboard works as expected. 

### Actual Behavior

No keyboard input works.

Terminal.Gui tracking issue: https://github.com/gui-cs/Terminal.Gui/issues/2745

**Comment by DHowett** [comment]
Hmm. The only meaningful keyboard-related changes I know that we made in Terminal 1.18 are:

1. We now emit key *up* events when a key is released rather than when it is pressed.
2. @lhecker may have changed something about input record generation?

I'd be interested to figure out exactly how Terminal.Gui is reading input using the .NET console APIs.

Before I dig into the project, I'm trying to boil it down to "this API call fails" and I've tried `System.Console.ReadKey(true)` but am not seeing anything unusual.

**Comment by DHowett** [comment]
_and 1.18_!

<img width="594" alt="image" src="https://github.com/microsoft/terminal/assets/189190/4d91539f-ba83-45ce-8bca-22e423d336e7">

**Comment by tig** [comment]
Well that's odd.

Just tested again to ensure I wasn't crazy.

Works fine in 1.17

In 1.18
<img width="625" alt="image" src="https://github.com/microsoft/terminal/assets/585482/1ab0f1ee-56db-4a10-9fe5-185da96b613a">

**Comment by zadjii-msft** [comment]
Can you try repro'ing this with the [debug tap](https://github.com/microsoft/terminal/wiki/Enabling-the-debug-tap) to get a trace of all the input and output? Once the bug starts occurring, send us a screenshot and we might be able to figure out what the Terminal thinks it's getting here.

(this might be too noisy for debug tap though. Pay attention to the red input sequences?)

**Comment by tig** [comment]
Found the cluprit. 

We have logic in `NetDriver` for when a terminal reports (via `CSI_ReportDeviceAttributes`) capabilities other than "VT101 with No Options". This logic is naïve, and was apparently not exercised previously. WT 1.17 used to emit “\x1b[?1;0c”, indicating "VT101 with No Options". It now emits "\x1b[?61;6;7;22;23;24;28;32;42c":

```cs
	/// <summary>
	/// The terminal reply to <see cref="CSI_ReportDeviceAttributes"/> :
	/// Windows Terminal v1.17 and below Will emit “\x1b[?1;0c”, indicating "VT101 with No Options".
	/// Windows Terminal v1.18+ emits: \x1b[?61;6;7;22;23;24;28;32;42c"
	/// - 61 - indicates VT525
	/// - 7 - indicates VT400
	/// - 6 - indicates Selective Erase
	/// - 22 - indicates ANSI Color
	/// - 23 - indicates ANSI Text Locator
	/// - 24 - indicates VT200 Highlighting
	/// - 28 - indicates Rectangular Editing
	/// - 32 - indicates National Replacement Character Sets
	/// - 42 - indicates ISO Latin-1 Character Set
	/// </summary>
	public const string CSI_ReportDeviceAttributes_Terminator = "c";
```

We assume if we see anything OTHER than "1;0c” we can use `CSI_ReportTerminalSizeInChars` to ask the terminal for its size. Which is probably ok. Except, we do it in every iteration of our mainloop which means all the terminal is doing is sending us size reports (via `Console.ReadKey`), and as a result no actual keypresses are getting through.

I think we can close this issue for WT because I don' think WT is doing anything wrong.

**Comment by tig** [comment]
> ```cs
> CSI_ReportDeviceAttributes
> ```

@zadjii-msft - Can you point in the right direction as to why WT is emitting what it is emitting in response to `CSI Ps c  Send Device Attributes`?

As noted above, I'm seeing "\x1b[?61;6;7;22;23;24;28;32;42c". But I can't find "61" documented anywhere, etc...

**Comment by zadjii-msft** [comment]
https://github.com/microsoft/terminal/pull/14906 is where we changed this

> This first parameter of the response is 61, representing a conformance level of 1. The subsequent parameters identify the supported feature extensions

**Comment by zadjii-msft** [comment]
@j4james  might be able to elaborate more. I'm not sure I totally understand what "conformance level of 1" means 😅

**Comment by DHowett** [comment]
Interesting. I can find reference to 62 meaning "service class 2"[^1] and 64 meaning "service class 4"[^2] for DA1.

[^1]: [VT220 Programming Manual](https://vt100.net/docs/vt220-rm/chapter4.html#S4.17.1), _"...I am a service class 2 terminal (62)..."_
[^2]: [VT510 Programming Manual](https://vt100.net/docs/vt510-rm/DA1.html), _"...determine the basic service class and level of the device. The VT510 is a level 4 terminal so its service class code is 64..."_

**Comment by j4james** [comment]
From the _DEVICE ATTRIBUTES (Primary)_ documentation in the _DEC STD 070 Video Systems Reference Manual_, page 4-17:

> The value of the first parameter is encoded so that a simple range check can be performed to determine the basic service class and level of the device. Each service class is designated a range of decimal values from 1 to 9. For example, Character Cell Display devices are in the range 61 to 69. The architectural level to which the device conforms is indicated by the low order digit. Thus, a response with a first parameter of 61 indicates that the device conforms to Level 1 of the Character Cell Display architecture.

If you want the exact definition of a level 1 device you'll need to read the manual, but it's basically the same capabilities as the VT100 range of terminals. So the point of our response is that we're around the same level as a VT102, but with a bunch of extensions. We don't yet meet the requirements for level 2.

**Comment by j4james** [comment]
> WT v1.18+ now emits "\x1b[?61;6;7;22;23;24;28;32;42c":
> 
> ```cs
> 	/// <summary>
> 	/// The terminal reply to <see cref="CSI_ReportDeviceAttributes"/> :
> 	/// Windows Terminal v1.17 and below Will emit “\x1b[?1;0c”, indicating "VT101 with No Options".
> 	/// Windows Terminal v1.18+ emits: \x1b[?61;6;7;22;23;24;28;32;42c"
> 	/// - 61 - indicates VT525
> 	/// - 7 - indicates VT400
> 	/// - 6 - indicates Selective Erase
> 	/// - 22 - indicates ANSI Color
> 	/// - 23 - indicates ANSI Text Locator
> 	/// - 24 - indicates VT200 Highlighting
> 	/// - 28 - indicates Rectangular Editing
> 	/// - 32 - indicates National Replacement Character Sets
> 	/// - 42 - indicates ISO Latin-1 Character Set
> 	/// </summary>
> ```

@tig In case you are planning on using this comment somewhere, you should be aware that the details listed here are almost entirely incorrect. If you run the _Primary Device Attributes_ test in [Vttest](https://www.invisible-island.net/vttest/) (test 6.4) you can see the correct interpretation of that response.

**Comment by tig** [comment]
Yep, those comments were generated by Copilot. I quickly realized it was on drugs. Here's how they read now:

```cs
/// <summary>
/// ESC [ 0 c - Send Device Attributes (Primary DA)
///
/// https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Application-Program-Command-functions
/// https://www.xfree86.org/current/ctlseqs.html
/// Windows Terminal v1.17 and below emits “\x1b[?1;0c”, indicating "VT101 with No Options".
/// Windows Terminal v1.18+ emits: \x1b[?61;6;7;22;23;24;28;32;42c"
/// See https://github.com/microsoft/terminal/pull/14906
///
/// 61 - The device conforms to level 1 of the character cell display architecture
/// (See https://github.com/microsoft/terminal/issues/15693#issuecomment-1633304497)
/// 6 = Selective erase
/// 7 = Soft fonts
/// 22 = Color text
/// 23 = Greek character sets
/// 24 = Turkish character sets
/// 28 = Rectangular area operations
/// 32 = Text macros
/// 42 = ISO Latin-2 character set
/// 
/// </summary>
```

**Comment by j4james** [comment]
@tig That's interesting. I'm kind of impressed that it came up with that, even if it is mostly wrong. Your updated comment looks good though.

**Comment by tig** [comment]
I love that all the AI folks are now using the term "hallucinations" to describe when they just make shit up.
